domain/modelupgrade/state/state.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/canonical/sqlair"
+
+	coredatabase "github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/semversion"
+	"github.com/juju/juju/domain"
+	modelerrors "github.com/juju/juju/domain/model/errors"
+	"github.com/juju/juju/internal/errors"
+)
+
+// State is used to access the database.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState creates a state to access the database.
+func NewState(factory coredatabase.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// GetModelVersionInfo returns the current model target version
+// and whether the model is the controller model or not.
+// The following errors can be expected:
+// - [modeleerrors.NotFound] when the model does not exist.
+// - [modelerrors.AgentVersionNotFound] when there is no target version found.
+func (st *State) GetModelVersionInfo(ctx context.Context) (semversion.Number, bool, error) {
+	db, err := st.DB()
+	if err != nil {
+		return semversion.Zero, false, errors.Capture(err)
+	}
+
+	res := dbAgentVersion{}
+	versionStmt, err := st.Prepare("SELECT &dbAgentVersion.* FROM agent_version", res)
```

Comment:
These are both single row tables, just get it all at once with a single unqualified join.